### DISCUSSION TITLE
Side module with different allocators

### DISCUSSION THREAD

**Discussion by technopedi** [emscripten-core/emscripten#20935] [Q&A]
Is it possible and advisable to have a side module with new and delete operators overloaded by another allocators like scalabale_malloc and scalable_free?

**Comment by sbc100** [comment]
IIRC there can only be one copy of the base new and delete in the whole program, but individual classes can override this.  I don't think it matter if you are using a side module or now, the limitations and best practices should be identical.    Unless  I'm misunderstanding something here.

**Reply by technopedi** [reply]
Incase I have an application, where i have a main module and a side module. If i override new and delete in main.cpp(belongs to main module) it works fne, but if i do same for only dll i.e. side module, it is not letting me to load dll itself saying  - main.js:631 Aborted(Assertion failed: undefined symbol '_Znwm'. perhaps a side module was not linked in? if this global was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment)

**Reply by sbc100** [reply]
That is a strange error, and it would be good to get the bottom of why you are seeing.  Can you share a failing example perhaps?

Having said that for any given symbol (e.g. new and delete) there can generally be only definition in the entire program.   Override a symbol in main module will also override it in the side module.  This is similar to how ELF works.. if multiple modules define the same symbol the first one the dyanmic linker sees will "win" and all other will be ignored.   In other words, you cannot have the same symbol defined by two modules.. one of them will always be ignored.

**Reply by technopedi** [reply]
So , even if dll wasn't loaded, dynamic linker will still not able to find new operator?

**Reply by technopedi** [reply]
if symbols are overridden inside module, it is safe to say that overridden symbols are used only in side module in emcc and not in main module code? i,e when side module code is called, then only the overridden symbols will be used, else previous symbols implementation will be used. A similar example - https://github.com/technopedi/sample-malloc

**Reply by sbc100** [reply]
My understanding is that symbols should only have one version, and since the main module is loaded first it should always be the main module wins the race to define a given symbol.
